src/Generator/Generators/CSharp/CSharpSources.cs

```diff
@@ -932,26 +939,40 @@ private void GenerateIndexerSetter(Function function)
             function.Type.IsPointerTo(out type);
             PrimitiveType primitiveType;
             var internalFunction = GetFunctionNativeIdentifier(function);
-            var @internal = $@"{Helpers.InternalStruct}{
-                Helpers.GetSuffixForInternal(function.Namespace)}";
+            TypePrinter.PushContext(TypePrinterContextKind.Native);
+            var @internal = function.Namespace.Visit(TypePrinter);
+            TypePrinter.PopContext();
+            var ctx = new CSharpMarshalContext(Context)
+            {
+                Parameter = new Parameter
+                {
+                    Name = "value",
+                    QualifiedType = new QualifiedType(type)
+                },
+                ReturnType = new QualifiedType(type)
+            };
+            var marshal = new CSharpMarshalManagedToNativePrinter(ctx);
+            type.CSharpMarshalToNative(marshal);
+            Write(marshal.Context.Before);
             if (type.IsPrimitiveType(out primitiveType))
             {
                 WriteLine($@"*{@internal}.{internalFunction}({
-                    GetInstanceParam(function)}, {function.Parameters[0].Name}) = value;");
+                    GetInstanceParam(function)}, {function.Parameters[0].Name}) = {
+                    marshal.Context.Return};");
             }
             else
             {
-                var typeString = type.ToString();
+                TypePrinter.PushContext(TypePrinterContextKind.Native);
```

Comment:
I see this pattern a couple times.

Let's abstract it into a static method, `PrintNativeTypeName`, or something.